### DISCUSSION TITLE
How many kafka topics will be created during work.

### DISCUSSION THREAD

**Discussion by wallyell** [milvus-io/milvus#26234] [Q&A and General discussion] [answered]
I found that there are 4 topics created in version 2.2.11 and 9 topics created in version 2.2.12 automatically:
by-dev-datacoord-timetick-channel
by-dev-rootcoord-delta_0
by-dev-rootcoord-delta_1
by-dev-rootcoord-delta_2
by-dev-rootcoord-delta_3
by-dev-rootcoord-dml_0
by-dev-rootcoord-dml_1
by-dev-rootcoord-dml_2
by-dev-rootcoord-dml_3

so how many kafka topics will be created during work and can I limit the number of kafka topics?

**Comment by yhmo** [comment, marked as answer]
No change between 2.2.11 and 2.2.12 for topics.
**by-dev-datacoord-timetick-channel** is the topic to pass time tick messages for every 200ms.

**by-dev-rootcoord-delta_0** is the topic to pass delete requests for collection_0
**by-dev-rootcoord-dml_0** is the topic to pass insert requests for collection_0

**by-dev-rootcoord-delta_N** is the topic to pass delete requests for collection_N
**by-dev-rootcoord-dml_N** is the topic to pass insert requests for collection_N

So, there must be N*2+1 topics in pulsar/kafka in both 2.2.11 and 2.2.12.

**Comment by xiaofan-luan** [comment]
this is bases on collections of you creating. but has a upper limit.
If you don't have enough collections, all the topic is lazily created

**Reply by shaiktas** [reply]
@xiaofan-luan great information, i was trying to set up a Java based KafkaConsumer to consume messages from above topics, but could not able to read message it looks some specific format. Can you pls suggest what should be consumer configs to consume messages from above topics? I am interested to see the messages being published by milvus in these topics?

**Reply by yhmo** [reply]
The message is encoded from structs, the structs are defined in milvus source code.
There is a tool [milvus-cdc](https://github.com/zilliztech/milvus-cdc) do the same thing with Golang. It is open-source, you can read its source code to know how the tool reads/parses messages from kafka/pulsar.
